add support for macOS

s-u opened the pull request:
This is a minimal set of code changes and additions to cmake rules to allow `flang` and `libpgmath` to build on macOS (aka Darwin).

[Tested using clang6 to build on Darwin 15.6.0 (aka OS X 10.11.6)]

gklimowicz commented:
Simon, can you do an individual contributor license agreement for us so we can take this pull request?

NVIDIA is requiring that we do CLAs for this sort of outside contribution.

This is described in the Flang wiki here: https://github.com/flang-compiler/flang/wiki/Community

gklimowicz commented:
@s-u FYI, I can't take these changes without a completed contributor license agreement. At least for about another few weeks.

If you have an update for this, please let me know.

kiranchandramohan commented:
@s-u CLA (contributor license agreement) is not a requirement now. If you are interested in submitting this PR please rebase, else consider closing.

s-u commented:
@kiranchandramohan thanks, rebased and updated.
I have tested the build using classic-flang-llvm-project (bootstrapped with Xcode 10.3) on macOS 10.14 (Mojave) = Darwin 18.

pawosm-arm commented:
Seems like each time I thought I was approving particular commit, it was approval for the entire PR!

pawosm-arm commented on runtime/flangrti/iostdinit.c:
```diff
@@ -153,7 +153,7 @@ __io_ferror(void *p)
 __io_getfd(void *fp)
 {
-  return (((FILE *)fp)->_fileno);
+  return fileno((FILE *)fp);
```
Good luck with that! I spotted this years before, and proposed change was rejected.

kiranchandramohan replied:
Any particular reason it was rejected? Does it apply now?

pawosm-arm replied:
AFAIR one of the building environments maintained by the original project owners did not provide `fileno()` function. A change like that was breaking compatibility with that platform.

kiranchandramohan replied:
@s-u Would it be possible to conditionally make this change for Mac?

s-u replied:
Obviously, I could, but why leave bad code in there? Accessing system's internal structure that are supposed to be opaque will break on any decent system, not just macOS (hint, hint, that's why the name starts with an underscore!).

kiranchandramohan replied:
OK. let us use your version. If we see a platform where there is an issue then we will change.

pawosm-arm reviewed (commented):
The commit message is about fileno(), yet the commit additionally sorts out dumpregs()/getRegs() definitions. This is considered as lame behavior, consider splitting this commit in two, or update the commit message to reflect correctly the scope of the change.

dnoan commented:
`runtime/flangrti/CMakeLists.txt` has `FLANG_LIBOMP` and `LIBPGMATH` hard-coded to `libomp.so` and `libpgmath.so`. These should be set to `libomp.dylib` and `libpgmath.dylib` on Mac. Otherwise the project doesn't compile.

s-u commented:
Build instruction are like in the Wiki only slightly modified - your mileage may wary depending on tools used.

```shell
BASE="`pwd`"
CMAKE=/Applications/CMake.app/Contents/bin/cmake

INSTALL_PREFIX=/opt/flang
CMAKE_OPTIONS="-DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX \
    -DLLVM_CONFIG=$INSTALL_PREFIX/bin/llvm-config \
    -DCMAKE_CXX_COMPILER=$INSTALL_PREFIX/bin/clang++ \
    -DCMAKE_C_COMPILER=$INSTALL_PREFIX/bin/clang \
    -DCMAKE_Fortran_COMPILER=$INSTALL_PREFIX/bin/flang 
    -DCMAKE_Fortran_COMPILER_ID=Flang \
    -DCMAKE_BUILD_TYPE=Release \
    -DLLVM_TARGETS_TO_BUILD=X86"

cd classic-flang-llvm-project
mkdir -p build && cd build
## There is a bug in that branch that uses wrong flags for xml2 so we have to disable it
$CMAKE $CMAKE_OPTIONS -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ \
       -DLLVM_ENABLE_CLASSIC_FLANG=ON -DLLVM_ENABLE_PROJECTS="clang;openmp;libcxx;libcxxabi" \
       -DLLVM_ENABLE_LIBXML2=OFF ../llvm
make -j12

cd "$BASE"

(cd flang/runtime/libpgmath
 mkdir -p build && cd build
 $CMAKE $CMAKE_OPTIONS ..
 make -j12
 make install)

(cd flang
 mkdir -p build && cd build
 $CMAKE $CMAKE_OPTIONS -DFLANG_LLVM_EXTENSIONS=ON \
	-DFLANG_LIBOMP=$INSTALL_PREFIX/lib/libomp.dylib -DLIBPGMATH=$INSTALL_PREFIX/lib/libpgmath.dylib ..
 make -j12
 make install)
```

@dnoan it says that those are undefined, so if you define them it works, no code change needed. I have now added a fix for the suffixes as well.
